--- php7.0.orig/Zend/zend_operators.h
+++ php7.0/Zend/zend_operators.h
@@ -700,9 +700,9 @@ static zend_always_inline void fast_long
 		"fstpl	(%0)\n"
 		"1:"
 		:
-		: "r"(&result->value),
-		  "r"(&op1->value),
-		  "r"(&op2->value),
+		: "g"(&result->value),
+		  "g"(&op1->value),
+		  "g"(&op2->value),
 		  "n"(IS_LONG),
 		  "n"(IS_DOUBLE),
 		  "n"(ZVAL_OFFSETOF_TYPE)
@@ -727,9 +727,9 @@ static zend_always_inline void fast_long
 		"fstpl	(%0)\n"
 		"1:"
 		:
-		: "r"(&result->value),
-		  "r"(&op1->value),
-		  "r"(&op2->value),
+		: "g"(&result->value),
+		  "g"(&op1->value),
+		  "g"(&op2->value),
 		  "n"(IS_LONG),
 		  "n"(IS_DOUBLE),
 		  "n"(ZVAL_OFFSETOF_TYPE)
@@ -757,9 +757,9 @@ static zend_always_inline void fast_long
 		"stfd 0, 0(%0)\n"
 		"1:"
 		:
-		: "r"(&result->value),
-		  "r"(&op1->value),
-		  "r"(&op2->value),
+		: "g"(&result->value),
+		  "g"(&op1->value),
+		  "g"(&op2->value),
 		  "n"(IS_LONG),
 		  "n"(IS_DOUBLE),
 		  "n"(ZVAL_OFFSETOF_TYPE)
