From: =?utf-8?q?Ond=C5=99ej_Sur=C3=BD?= <ondrej@sury.org>
Date: Tue, 1 Dec 2015 11:34:19 +0100
Subject: Include php_config.h if we don't have config.h to get right includes

---
 ext/gd/gd_compat.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/ext/gd/gd_compat.c b/ext/gd/gd_compat.c
index a882efa..44a4910 100644
--- a/ext/gd/gd_compat.c
+++ b/ext/gd/gd_compat.c
@@ -1,5 +1,7 @@
 #ifdef HAVE_CONFIG_H
 #include "config.h"
+#else
+#include "php_config.h"
 #endif
 
 #ifdef HAVE_GD_PNG
